### PR TITLE
Add error if differentiating diff_method=None

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

If someone tries to differentiate a qnode that has `diff_method=None`, they may get less-than-useful error message.  We also assume that no interface parameters are used with `diff_method=None`

**Description of the Change:**

Uses the ML boundary when `diff_method=None` and raises an informative error if differentiation is requested.

**Benefits:**

ML parameters can be used with `diff_method=None`, and incorrect usage provides good error messages.

**Possible Drawbacks:**

**Related GitHub Issues:**

Solves #6769 [sc-81490]

### CODE REVIEW

**Comment by PietropaoloFrisoni** [comment]
@albi3ro Should this be part of the release?

**Comment by albi3ro** [comment]
> @albi3ro Should this be part of the release?

Nope. It's changing architectural components and core behavior.

File: tests/workflow/interfaces/test_jacobian_products.py [hunk 1]
```diff
@@ -79,6 +80,32 @@ def _tol_for_shots(shots):
...
+
+    with pytest.raises(
+        qml.QuantumFunctionError, match="cannot be calcualted with diff_method=None"
+    ):
+        jpc.execute_and_compute_jvp((), ())
+
+    with pytest.raises(
+        qml.QuantumFunctionError, match="cannot be calcualted with diff_method=None"
+    ):
+        jpc.execute_and_compute_jacobian(())
```
Comment (albi3ro) [inline tests/workflow/interfaces/test_jacobian_products.py:106]:
```suggestion
    with pytest.raises(
        qml.QuantumFunctionError, match="cannot be calculated with diff_method=None"
    ):
        jpc.compute_jacobian(())

    with pytest.raises(
        qml.QuantumFunctionError, match="cannot be calculated with diff_method=None"
    ):
        jpc.compute_vjp((), ())

    with pytest.raises(
        qml.QuantumFunctionError, match="cannot be calculated with diff_method=None"
    ):
        jpc.execute_and_compute_jvp((), ())

    with pytest.raises(
        qml.QuantumFunctionError, match="cannot be calculated with diff_method=None"
    ):
        jpc.execute_and_compute_jacobian(())
```

File: tests/gradients/core/test_pulse_gradient.py
Comment (albi3ro) [inline tests/gradients/core/test_pulse_gradient.py:1]: `qml.execute` is now properly setting the trainable params on the input tapes.

**Review by andrijapau — APPROVED** [review]
This is great 🚀 !

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -15,6 +15,9 @@
 * The coefficients of observables now have improved differentiability.
   [(#6598)](https://github.com/PennyLaneAI/pennylane/pull/6598)
 
+* An informative error is raised in a qnode with `diff_method=None` is differentiated.
```
Comment (JerryChen97) [inline doc/releases/changelog-dev.md:18]:
```suggestion
* An informative error is raised when a `QNode` with `diff_method=None` is differentiated.
```

File: pennylane/workflow/jacobian_products.py [hunk 3]
```diff
@@ -210,6 +210,24 @@ def execute_and_compute_jacobian(self, tapes: QuantumScriptBatch) -> tuple[Resul
...
+
+    def execute_and_compute_jvp(
+        self, tapes: QuantumScriptBatch, tangents: Sequence[Sequence[TensorLike]]
+    ) -> tuple[ResultBatch, tuple]:
+        raise qml.QuantumFunctionError("Derivatives cannot be calculated with diff_method=None")
+
+    def execute_and_compute_jacobian(self, tapes: QuantumScriptBatch) -> tuple[ResultBatch, tuple]:
+        raise qml.QuantumFunctionError("Derivatives cannot be calculated with diff_method=None")
+
+
```
Comment (JerryChen97) [inline pennylane/workflow/jacobian_products.py:230]:
```suggestion
class NoGradients(JacobianProductCalculator):
    """A jacobian product calculator that raises errors when a vjp or jvp is requested."""
    
    error_msg = "Derivatives cannot be calculated with diff_method=None"

    def compute_jacobian(self, tapes: QuantumScriptBatch) -> tuple:
        raise qml.QuantumFunctionError(NoGradients.error_msg)

    def compute_vjp(self, tapes: QuantumScriptBatch, dy: Sequence[Sequence[TensorLike]]) -> tuple:
        raise qml.QuantumFunctionError(NoGradients.error_msg)

    def execute_and_compute_jvp(
        self, tapes: QuantumScriptBatch, tangents: Sequence[Sequence[TensorLike]]
    ) -> tuple[ResultBatch, tuple]:
        raise qml.QuantumFunctionError(NoGradients.error_msg)

    def execute_and_compute_jacobian(self, tapes: QuantumScriptBatch) -> tuple[ResultBatch, tuple]:
        raise qml.QuantumFunctionError(NoGradients.error_msg)


```
Not very important but since msg are the same we should manage them together instead of hardcode

File: tests/gradients/core/test_pulse_gradient.py [hunk 4]
```diff
@@ -1149,6 +1150,9 @@ def test_constant_commuting(self, t):
             -2 * jnp.sin(2 * p[1]) * jnp.cos(2 * p[0]) * (T[1] - T[0]),
         ]
         assert qml.math.allclose(res, exp_grad)
+        r = qml.execute([tape], dev, None)
+        # Effective rotation parameters
+        assert qml.math.isclose(r, jnp.cos(2 * p[0]) * jnp.cos(2 * p[1]))
```
Comment (JerryChen97) [inline tests/gradients/core/test_pulse_gradient.py:1155]:
```suggestion
        exp = jnp.cos(2 * p[0]) * jnp.cos(2 * p[1])
        assert qml.math.isclose(r, exp)
```

**Review by JerryChen97 — APPROVED** [review]
LGTM! Just some good-to-have suggestions
